Openworld: Introduce itemspec_t struct (with associated helpers and tests) and implement probability-based item generation

Justin08784 opened the pull request:
- Introduced `itemspec_t` struct, which allows game authors to –– per `roomspec_t` –– customize item occurrence probability and possible quantity ranges for the item. Relevant tests for helpers–– new, init, and free–– have been added to test_structs.c
The hope is to be able to generate a more diverse and exciting map. 
- Created a new function, generate_items(), which generates the items for each roomspec based on its itemspecs. If an item does not have a corresponding itemspec, the function resorts to the default behavior of spawning with probability 1 (guaranteed spawn) and quantity 1. Relevant tests have been added to test_autogenerate.c
- roomspec_to_room() now calls generate_items() instead of random_items(), which just picked a random subset of items from the roomspec.
- Refactored copy_item_to_hash() so that it can add duplicate items to the same room–– see last year's [item-stacking implementation](https://github.com/uchicago-cs/chiventure/issues/703)

ahmed-sumaiya reviewed (approved):
Looks good to me!

MaxineK36 commented on src/openworld/src/autogenerate.c:
```diff
@@ -214,6 +214,54 @@ int random_item_lookup(item_hash_t **dst, item_hash_t *src, int num_iters)
+        if (total_count == MAX_RAND_ITEMS)
+            break;
+
+        double spawn_chance = 1;
```
is there a reason these are given initial default values of 1?

Justin08784 replied:
To mimic the behavior of random_items (the previous function that was called by roomspec_to_room).

random_items spawns 1 instance of each item specified in the roomspec item hash. However, the behavior is slightly different in that random_items() picks a random subset of the item hash, while generate_items iterates through the item hash in a consistent manner.

MaxineK36 replied:
Ok, makes sense. Please note this in a comment in this section of code.

MaxineK36 replied:
Thanks for adding the comment! For stylistic consistency, please put the block comment *before* the relevant section of code (instead of indicating with "^^")

MaxineK36 commented on src/openworld/src/autogenerate.c:
```diff
@@ -214,6 +214,54 @@ int random_item_lookup(item_hash_t **dst, item_hash_t *src, int num_iters)
+            max_num = itemspec->max_num;
+            min_num = itemspec->min_num;
+        }
+        int num_quantities = max_num - min_num + 1;
```
what's the logic behind this calculation?

Justin08784 replied:
min_num and max_num prescribes the interval of possible quantities in which the item can spawn. To compute the number of quantities (we need this for calculating the probability of each quantity), we do max - min + 1.

e.g. [2, 4] has three possible values of 2, 3, and 4. This is computed by 4 - 2 + 1 .
Since we decided each quantity is equally likely to occur, the probability is 1/3 = 0.33...

MaxineK36 replied:
Ok, got it. Please add comments next to `min_num` and `max_num` clarifying that these refer to possible quantities the item can spawn.

MaxineK36 replied:
Same note here—comment looks great, but please move it up

MaxineK36 commented on src/openworld/src/autogenerate.c:
```diff
@@ -214,6 +214,54 @@ int random_item_lookup(item_hash_t **dst, item_hash_t *src, int num_iters)
+        int num_quantities = max_num - min_num + 1;
+
+        int spawn_num = min_num;
+        bool should_spawn = ((double) rand()) / RAND_MAX <= spawn_chance;
```
why is this separated as a bool instead of placed directly in the if?

Justin08784 replied:
I thought it might look less cluttered.

MaxineK36 replied:
In this case, I think it'll be more readable in the `if`; declaring a separate bool value usually indicates that it'll be used elsewhere

MaxineK36 commented on src/openworld/src/autogenerate.c:
```diff
@@ -214,6 +214,54 @@ int random_item_lookup(item_hash_t **dst, item_hash_t *src, int num_iters)
+
+    int total_count = 0;
+    item_t *curr, *tmp;
+    item_hash_t *rv = NULL;
```
Why is this called `rv`?

Justin08784 replied:
rv for return value

MaxineK36 replied:
Can you name it something more descriptive?

MaxineK36 reviewed (commented):
A couple of clarification questions, but looks good overall!

MaxineK36 commented on src/openworld/src/autogenerate.c:
```diff
@@ -214,6 +214,66 @@ int random_item_lookup(item_hash_t **dst, item_hash_t *src, int num_iters)
+        */
+
+        int spawn_num = min_num;
+        if (((double) rand()) / RAND_MAX <= spawn_chance) {
```
Please add parens around `((double) rand()) / RAND_MAX`

MaxineK36 reviewed (changes requested):
Just a few stylistic fixes!

MaxineK36 reviewed (approved):
LGTM!

Justin08784 commented:
Closing statement:
Comments by reviewers have been addressed and branch `openworld/add-itemspec` is now merged into dev.

_Summary:_
Additions
- Added itemspec_t struct for specification of occurrence probability and quantity ranges for an item in a specific roomspec. 
- Created new function, generate_items(), for itemspec-based item generation.

_Changes_  
- roomspec_to_room() calls generate_items() instead of random_items(), which just picked a random subset from the roomspec item hash.
- Refactored copy_item_to_hash() so that it can add duplicate items to the same room using item stacking

MaxineK36 commented:
Issue Score: ✔️++

Comments:
Awesome work! Your closing statement is particularly clear and well-written.

<!-- {"score": "check-plus-plus"} -->
